packages/cli-node/src/pacman/PackageManager.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Yarn } from './yarn';
+import { Lockfile } from './Lockfile';
+import { SpawnOptionsPartialEnv } from '../run';
+import { paths } from '../paths';
+import fs from 'fs-extra';
+
+/**
+ * Package info retrieved from the package manager, usually from NPM.
+ *
+ * @public
+ */
+export type PackageInfo = {
+  name: string;
+  'dist-tags': Record<string, string>;
+  versions: string[];
+  time: { [version: string]: string };
+};
+
+/**
+ * Represents the package manager in use by this instance of Backstage. This
+ * interface allows Backstage adopters to change the package manager used by
+ * their repo and still use the Backstage CLI, and it's helpful tooling.
+ *
+ * @public
+ */
+export interface PackageManager {
+  /** The name of the package manager. */
+  name(): string;
+
+  /** The self-reported version of the package manager. */
+  version(): string;
+
+  /** The file name of the lockfile used by the package manager. */
+  lockfileName(): string;
+
+  /**
+   * If this repo is a monorepo, returns the patterns specified by the package
+   * manager's monorepo configuration. Does not attempt to resolve any globs.
+   */
+  getMonorepoPackages(): Promise<string[]>;
```

Comment:
Thanks for bringing up this use case. How does it currently work for you with Yarn? Do you have a `yarn.lock` and `package.json` (with the `workspaces` field) at the root of your Backstage instance (rather than the monorepo root) and does `backstage-cli` correctly locate these files?

In my test branch with PNPM (not pushed yet) I've been looking to replicate the same behaviour, but it sounds like we might need additional configuration in your case since the PNPM files won't be located in the same place the Yarn files are.

Reply:
We run the base Yarn setup in a separate monorepo, this work would help us move to an existing Rush monorepo. Maybe this would fit into `backstage.json`? Rush has this idea of a packageManager field in `rush.json`, https://rushjs.io/pages/maintainer/package_managers/#specifying-your-package-manager. And we could allow additional config for where your lockfile is etc

Reply:
Very interesting. Rush looks pretty cool and I had not heard of it so thanks again for bringing this up. I'm not sure what the right path forward here is, it almost seems like Rush could be it's own implementation of `PackageManager`/`Lockfile` but then again it still uses Yarn/pnpm/npm under the hood so there'd be a lot of shared code. Probably the simplest option would be to support Rush similarly to how lerna is supported, with a few branching conditions to retrieve files from the right locations where needed. Unfortunately I don't think it's a good idea to include that scope in these changes right off the bat, but it will certainly be much easier to do once we have the base Pnpm support merged.

Reply:
Happy to have this be handled later - just wanted to bring it up a little early. Basically, the biggest thing for me is that I'd love to see package manager be definable and lockfile location be customizable, yarn/npm/pnpm. I can run with the logic from there. A lot of libraries try the autodiscovery method and it falls down for Rush 😕.

Reply:
rush "looks" cool, but it's not very useful. I feel that it lacks behind in features compared to a plain pnpm monorepo and slows us down more than anything. I've had incomparably better experience with pnpm+turbo+beachball (or changesets). Rush is a jack of all trades, master of none.